New TorusKnotBufferGeometry

Mugen87 opened the pull request:
This PR introduces `TorusKnotBufferGeometry`. Here is the corresponding [test fiddle](https://jsfiddle.net/fuwpLv10/7/).

There are some things i want to discuss:
- I'm afraid that the parameter `radialSegments` and `tubularSegments` have interchanged meanings in `TorusGeometry` and `TorusKnotGeometry`. I suggest that the implementation of `TorusGeometry` is correct, where `tubularSegments` determines the amount of vertical segments of the tube and `radialSegments` the amount of horizontal segments. Maybe it's good to align this in `TorusKnotBufferGeometry`.
- The `heightScale` parameter is confusing. It scales the z-component of a position on the torus curve which is quite extraordinary. Other geometries of three.js don't have such a parameter. And why is there no `widthScale` and `depthScale`? This implementation is somehow inconsistent. Is it basically possible to remove `heightScale` or mark it as deprecated?

Besides, i've added the `arc` parameter so `TorusKnotBufferGeometry` and  `TorusBufferGeometry` have more similar interfaces.

Mugen87 committed:
New TorusKnotBufferGeometry

WestLangley commented:
>I'm afraid that the parameter radialSegments and tubularSegments have interchanged meaning

Oops. Also look at `THREE.TubeGeometry`. And `THREE.CylinderGeometry` for that matter. There have been a lot of cooks in the kitchen. : - )

>Is it basically possible to remove heightScale or mark it as deprecated?

Yes, deprecate with a warning. That is what `BufferGeometry.scale()` is for.

>i've added the arc parameter so TorusKnotBufferGeometry and TorusBufferGeometry have more similar interfaces.

I do not think a torus knot needs an arc parameter.

Mugen87 commented:
The last commit removes the new `arc` parameter and marks `heightScale` as deprecated (with warning). Besides, i changed the meanings `radialSegments` and `tubularSegments` and adjusted the corresponding examples and docs.

WestLangley commented:
Changing the variable names is one thing, but you are changing the constructor signatures. This will cause a lot of breakage....

Mugen87 commented:
I see. But that means, the interfaces of `TorusKnotBufferGeometry` and `TorusBufferGeometry` will differ. Anyway, i will undone the last commit and just rename the variables...

Mugen87 committed:
TorusKnotBufferGeometry: Parameter changes

Mugen87 commented on src/loaders/ObjectLoader.js:
```diff
@@ -225,15 +225,15 @@ THREE.ObjectLoader.prototype = {
 							data.radius,
 							data.tube,
 							data.radialSegments,
 							data.tubularSegments,
```
How should we handle this case? If we change `data.radialSegments` and `data.tubularSegments` things could break again, right?

WestLangley replied:
Just do what you think is the least disruptive. There are no right answers here.

WestLangley commented:
Unfortunately, this is a mess. I wonder if we could use only these terms: `segments` along the path or tube, `radialSegments` around the tube circumference. Just a thought.

Mugen87 committed:
TorusKnotBufferGeometry: fix loader

Mugen87 commented:
@WestLangley What do you think is the best way to document such conventions? Is there a place in the wiki for this (something like "coding conventions")?

Mugen87 commented:
I've added a vertexNormalHelper ot the [fiddle](https://jsfiddle.net/fuwpLv10/7/). Now i have to think at a hedgehog when i see the many lines on the torus :smile:

mrdoob commented on src/extras/geometries/TorusKnotBufferGeometry.js:
```diff
@@ -0,0 +1,166 @@
+		q: q
+	};
+
+	if( heightScale !== undefined ) console.warn( 'THREE.TorusKnotGeometry: heightScale has been deprecated. Use .scale( x, y, z ) instead.' );
```
Considering that this is a new class that no one has used yet we can ignore `heightScale`. We should definitely keep the check in `TorusKnowGeometry` though.

Mugen87 replied:
Ok! I've just implemented your feedback.

Mugen87 committed:
TorusKnotBufferGeometry: Removed heightScale

mrdoob commented:
Many thanks!
